addons/threadsafety.py

```diff
@@ -1,39 +1,390 @@
+                        'warning',
+                        'Local constant static '
+                        + vartype + "'" + var.nameToken.str
+                        + "', dangerous if it is initialized"
+                        + ' in parallel threads',
+                        'threadsafety-const')
             else:
-                reportError(var.typeStartToken, 'warning', 'Local static ' + type + ': ' + var.nameToken.str, 'threadsafety')
+                reportError(var.typeStartToken, 'warning',
+                            'Local static ' + vartype + ': '
+                            + var.nameToken.str,
+                            'threadsafety')
+
 
-for arg in sys.argv[1:]:
-    if arg.startswith('-'):
-        continue
+def check_MTunsafe(data, srcfile, quiet=False):
+    """
+    Look for functions marked MT-unsafe in their man pages.
 
-    print('Checking %s...' % arg)
-    data = cppcheckdata.CppcheckData(arg)
+    The MT-unsafe functions are listed in id_MTunsafe (and id_MTunsafe_full).
+    That section of code can be regenerated by the external script MT-Unsafe.py
+    """
+    # Assume that the code is Multi-thread safe until proven otherwise
+    MTsafe = True
 
+    # go through each configuration
     for cfg in data.iterconfigurations():
-        print('Checking %s, config %s...' % (arg, cfg.name))
-        checkstatic(cfg)
+        if not quiet:
+            print('Checking %s, config %s...' % (srcfile, cfg.name))
+
+        # go through all tokens
+        for token in cfg.tokenlist:
+            if token.str in id_MTunsafe:
+                cppcheckdata.reportError(token, 'warning',
+                                         token.str + ' is MT-unsafe',
+                                         'threadsafety',
+                                         'unsafe-call')
+                MTsafe = False
+
+    return MTsafe
+
+
+def get_args_parser():    # noqa: D103
+    parser = cppcheckdata.ArgumentParser()
+    return parser
+
+
+if __name__ == '__main__':
+    parser = get_args_parser()
+    args = parser.parse_args()
+
+    exit_code = 0
+    quiet = not any((args.quiet, args.cli))
+
+    if not args.dumpfile:
+        if not args.quiet:
+            print('no input files.')
+        sys.exit(0)
+
+    for dumpfile in args.dumpfile:
+        # for arg in sys.argv[1:]:
+        #    if arg.startswith('-'):
+        #        continue
+
+        if not args.quiet:
+            print('Checking ' + dumpfile + '...')
+
+        # load XML from .dump file
+        data = cppcheckdata.CppcheckData(dumpfile)
+        # data = cppcheckdata.CppcheckData(args)
+
+        # Convert dump file path to source file in format generated by
+        # cppcheck. For example after the following call:
+        #    cppcheck ./src/my-src.c --dump
+        # We got 'src/my-src.c' value for 'file' field in cppcheckdata.
+        srcfile = dumpfile.rstrip('.dump')
+        srcfile = os.path.expanduser(srcfile)
```

Comment:
hmm.. this conversion does not seem to work if a cppcheck build dir is used.

The top of a dump file should look something like this:

```
<?xml version="1.0"?>
<dumps language="c">
  <platform name="native" char_bit="8" short_bit="16" int_bit="32" long_bit="64" long_long_bit="64" pointer_bit="64"/>
  <rawtokens>
    <file index="0" name="1.c"/>
```

Spontanously I think it's better to read `srcfile` from the `<file ..>` element that has `index="0"`.

How you access this information in the easiest way from cppcheckdata.py I don't know off the top of my head.